### PR TITLE
mm_realloc first try

### PR DESCRIPTION
**Pull request by vasilyrud**
[https://media.giphy.com/media/hdEhU942MSM6Y/giphy.gif](https://media.giphy.com/media/hdEhU942MSM6Y/giphy.gif)

mm_realloc

@quanvuong

### CODE REVIEW

File: mm.c [hunk 1]
```diff
@@ -523,19 +523,74 @@ void mm_free(void *ptr)
...
+	// adjust to be at start of block
+	char **old = (char **)ptr - 1;
+	char **bp = (char **)ptr - 1;
+
+	// get intended and current size
+	size_t new_size = ALIGN(size) / WORD_SIZE; // in words
+	size_t old_size = GET_SIZE(bp); // in words
+
+	// equivalent to mm_free if size == 0
+	if (new_size == 0) {
```
Comment (quanvuong) [inline mm.c:591]: should put this edge case before jumping into the meat of the function

File: mm.c [hunk 2]
```diff
@@ -523,19 +523,74 @@ void mm_free(void *ptr)
...
+		mm_free(ptr);
+		return NULL;
+	} else if (new_size < old_size) { // checks if can avoid reallocating
+		PUT_WORD(bp, PACK(old_size, FREE));
+    PUT_WORD(FTRP(bp), PACK(old_size, FREE));
+		alloc_free_block(bp, new_size);
```
Comment (quanvuong) [inline mm.c]: wouldn't this lead to external fragmentation as bp+(old_size - new_size) to bp+old_size free space, but not in free block ?

File: mm.c [hunk 3]
```diff
@@ -523,19 +523,74 @@ void mm_free(void *ptr)
...
+	} else if (new_size > old_size) {
+		if (GET_SIZE(NEXT_BLOCK_IN_HEAP(bp)) + old_size + 2 >= new_size &&
```
Comment (quanvuong) [inline mm.c]: use MACRO for constant

**Comment by quanvuong** [comment]
@vasilyrud this gives 97/100 

http://giphy.com/gifs/thedailyshow-yes-the-daily-show-yay-xT9DPDp0PNRFbuqqli

**Comment by vasilyrud** [comment]
Looks good to me.

Let's merge.

**Pull request merged**